base/tps/src/main/java/org/dogtagpki/server/tps/rest/base/TPSCertProcessor.java

```diff
@@ -0,0 +1,196 @@
+//
+// Copyright Red Hat, Inc.
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+//
+package org.dogtagpki.server.tps.rest.base;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.dogtagpki.server.rest.v2.PKIServlet;
+import org.dogtagpki.server.tps.TPSEngine;
+import org.dogtagpki.server.tps.TPSSubsystem;
+import org.dogtagpki.server.tps.dbs.TPSCertDatabase;
+import org.dogtagpki.server.tps.dbs.TPSCertRecord;
+import org.dogtagpki.server.tps.dbs.TokenDatabase;
+import org.dogtagpki.server.tps.dbs.TokenRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netscape.certsrv.base.BadRequestException;
+import com.netscape.certsrv.base.PKIException;
+import com.netscape.certsrv.tps.cert.TPSCertCollection;
+import com.netscape.certsrv.tps.cert.TPSCertData;
+import com.netscape.certsrv.user.UserResource;
+
+/**
+ * @author Marco Fargetta {@literal <mfargett@redhat.com>}
+ * @author Endi S. Dewata
+ */
+public class TPSCertProcessor {
+    private static final Logger logger = LoggerFactory.getLogger(TPSCertProcessor.class);
+    private TPSSubsystem subsystem;
+
+
+    public TPSCertProcessor(TPSEngine engine) {
+        subsystem = (TPSSubsystem) engine.getSubsystem(TPSSubsystem.ID);
+    }
+
+
+    public TPSCertCollection findCerts(List<String> authorizedProfiles, String tokenID, String filter, int start, int size) {
+        String method = "TPSCertProcessor:findCerts: ";
+        String msg = "";
+        logger.debug("TPSCertProcessor.findCerts({}, {}, {}, {})", filter, tokenID, start, size);
+
+        if (filter != null && filter.length() < PKIServlet.MIN_FILTER_LENGTH) {
+            throw new BadRequestException("Filter is too short. Must be at least " + PKIServlet.MIN_FILTER_LENGTH + " characters.");
+        }
+
+        if(tokenID == null) {
+            return findAllCerts(authorizedProfiles, filter, start, size);
+        }
+
+        Map<String, String> attributes = new HashMap<>();
+        if (tokenID != null) {
+            attributes.put("tokenID", tokenID);
+        }
```

Comment:
IIRC the problem is that the Certs page in TPS Web UI will pull all cert records in order to count the total number of entries so the UI can display the paging links such as **Page 1**, **Page 2**, ... **Page N**. If there are a lot of certs, this page will take a long time to open since by default there's no filter to narrow down the search, so that's why the minimum filter length was added, or if the token ID is available it can also be used to narrow down the search.

I think ideally we should change the UI to no longer require knowing the total entries, so the paging link we need to provide is just **Next Page** as long as there are more certs to display, then we don't have to require a minimum filter anymore.

Reply:
OK, I have update the code to match current implementation and added a TODO to fix this behaviour when the UI will be fixed